File: hdf-converters/shared/go/exportmap/exportmap_test.go
package exportmap

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestFloatToken(t *testing.T) {
	assert.Equal(t, json.Number("0.0"), FloatToken(0))
	assert.Equal(t, json.Number("7.5"), FloatToken(7.5))
	assert.Equal(t, json.Number("8.0"), FloatToken(8))
	assert.Equal(t, json.Number("10.0"), FloatToken(10))
	assert.Equal(t, json.Number("9.8"), FloatToken(9.8))
	// renders as a bare decimal token (not a quoted string, not an integer)
	line, err := EncodeLine(map[string]interface{}{"base_score": FloatToken(10)})
	require.NoError(t, err)
	assert.Equal(t, "{\"base_score\":10.0}\n", string(line))
}

func mkResults(statuses ...string) map[string]interface{} {
	var results []interface{}
	for _, s := range statuses {
		results = append(results, map[string]interface{}{"status": s})
	}
	return map[string]interface{}{"results": results}
}

func TestWorstOfResults(t *testing.T) {
	assert.Equal(t, "failed", WorstOfResults(mkResults("passed", "failed")))
	assert.Equal(t, "passed", WorstOfResults(mkResults("passed", "passed")))
	assert.Equal(t, "error", WorstOfResults(mkResults("error", "passed")))
	assert.Equal(t, "notApplicable", WorstOfResults(mkResults("notApplicable")))
	assert.Equal(t, "notReviewed", WorstOfResults(mkResults()))
}

func TestStatusOf(t *testing.T) {
	// no override: rollup == raw, not overridden, not suppressed
	st := StatusOf(mkResults("failed"))
	assert.Equal(t, "failed", st.Raw)
	assert.Equal(t, "failed", st.Rollup)
	assert.False(t, st.Overridden)
	assert.False(t, st.Suppressed)

	// Stored effectiveStatus is an output cache and is never read: with no
	// override records the rollup stays the results roll-up, nothing is
	// overridden or suppressed.
	req := mkResults("failed")
	req["effectiveStatus"] = "passed"
	st = StatusOf(req)
	assert.Equal(t, "failed", st.Raw, "raw stays the lossless results roll-up")
	assert.Equal(t, "failed", st.Rollup, "stale stored value ignored")
	assert.False(t, st.Overridden)
	assert.False(t, st.Suppressed)

	// A governing (non-expired) override drives the rollup: overridden true and,
	// since raw is failing while the effective status is not, suppressed true.
	req2 := mkResults("failed")
	req2["statusOverrides"] = []interface{}{map[string]interface{}{
		"type": "waiver", "status": "passed",
		"appliedAt": "2026-01-02T00:00:00Z", "expiresAt": "2099-12-31T00:00:00Z",
	}}
	st = StatusOf(req2)
	assert.True(t, st.Overridden)
	assert.Equal(t, "passed", st.Rollup)
	assert.True(t, st.Suppressed, "raw-failing driven non-failing → suppressed")

	// statusOverrides present but none governing (statusless): overridden true,
	// rollup == raw, NOT suppressed.
	req3 := mkResults("failed")
	req3["statusOverrides"] = []interface{}{map[string]interface{}{"type": "waiver"}}
	st = StatusOf(req3)
	assert.True(t, st.Overridden)
	assert.Equal(t, "failed", st.Rollup)
	assert.False(t, st.Suppressed)
}

// TestStatusOf_SuppressedMatrix pins the acceptance axis across dispositions:
// suppressing overrides (waiver/falsePositive/attestation drive effectiveStatus
// non-failing) → Suppressed; risk-response overrides that keep the finding
// failing (riskAdjustment/operationalRequirement/poam) → NOT suppressed.
func TestStatusOf_SuppressedMatrix(t *testing.T) {
	// A governing override driving a failing raw verdict to a non-failing
	// effective status suppresses (the verdicts a waiver/falsePositive yields).
	governing := func(status string) []interface{} {
		return []interface{}{map[string]interface{}{
			"type": "waiver", "status": status,
			"appliedAt": "2026-01-02T00:00:00Z", "expiresAt": "2099-12-31T00:00:00Z",
		}}
	}
	for _, eff := range []string{"passed", "notApplicable"} {
		req := mkResults("failed")
		req["statusOverrides"] = governing(eff)
		assert.True(t, StatusOf(req).Suppressed, "raw failed + governing override %s → suppressed", eff)
	}

	// riskAdjustment / operationalRequirement / poam: no status override — the
	// effective status stays failed and the finding stays actionable.
	req := mkResults("failed")
	req["statusOverrides"] = []interface{}{map[string]interface{}{"type": "riskAdjustment", "impact": map[string]interface{}{"value": 0.2}}}
	st := StatusOf(req)
	assert.True(t, st.Overridden)
	assert.False(t, st.Suppressed, "risk-adjusted failure stays actionable")

	// structural impact-0 notApplicable is NOT acceptance-suppression
	na := mkResults("failed")
	na["impact"] = float64(0)
	assert.False(t, StatusOf(na).Suppressed)
	assert.Equal(t, "notApplicable", StatusOf(na).Rollup)

	// a passing finding is never suppressed regardless of overrides
	pass := mkResults("passed")
	pass["statusOverrides"] = governing("passed")
	assert.False(t, StatusOf(pass).Suppressed)

	// an errored finding driven to passed is not suppressed (error isn't failing)
	erf := mkResults("error")
	erf["statusOverrides"] = governing("passed")
	assert.False(t, StatusOf(erf).Suppressed)

	assert.True(t, IsFailing("failed"))
	assert.False(t, IsFailing("error"))
	assert.False(t, IsFailing("passed"))
}

func TestGenericAccess(t *testing.T) {
	m, ok := AsMap(map[string]interface{}{"a": 1})
	assert.True(t, ok)
	assert.Equal(t, "", GetStr(m, "a")) // non-string -> ""
	assert.Equal(t, "", GetStr(nil, "x"))

	_, ok = AsMap([]interface{}{})
	assert.False(t, ok)
	s, ok := AsSlice([]interface{}{1, 2})
	assert.True(t, ok)
	assert.Len(t, s, 2)

	dst := map[string]interface{}{}
	SetIf(dst, "keep", "v")
	SetIf(dst, "drop", "")
	assert.Equal(t, map[string]interface{}{"keep": "v"}, dst)

	assert.Equal(t, []string{"x"}, StringSlice("x"))
	assert.Equal(t, []string{"a", "b"}, StringSlice([]interface{}{"a", 1, "b"}))
	assert.Nil(t, StringSlice(42))
}

func TestExtraction(t *testing.T) {
	doc := map[string]interface{}{"components": []interface{}{map[string]interface{}{"name": "web"}}}
	assert.Equal(t, "web", GetStr(FirstComponent(doc), "name"))
	assert.Nil(t, FirstComponent(map[string]interface{}{}))

	req := map[string]interface{}{
		"results":      []interface{}{map[string]interface{}{"startTime": "2024-01-01T00:00:00Z"}},
		"descriptions": []interface{}{map[string]interface{}{"label": "default", "data": "desc"}},
		"refs":         []interface{}{map[string]interface{}{"url": "https://x"}},
	}
	assert.Equal(t, "2024-01-01T00:00:00Z", FirstResultStartTime(req, "fb"))
	assert.Equal(t, "fb", FirstResultStartTime(map[string]interface{}{}, "fb"))
	assert.Equal(t, "desc", DefaultDescription(req))
	assert.Equal(t, "https://x", FirstRefURL(req))

	comp := map[string]interface{}{"componentId": "c1"}
	assert.Equal(t, "c1|base|V-1", EventID(comp, "base", "V-1"))
	assert.Equal(t, "|base|V-1", EventID(nil, "base", "V-1"))
}

func TestEncodeLine(t *testing.T) {
	// key-sorted, HTML not escaped, trailing newline
	line, err := EncodeLine(map[string]interface{}{"b": 1, "a": "x&y"})
	require.NoError(t, err)
	assert.Equal(t, "{\"a\":\"x&y\",\"b\":1}\n", string(line))
}

func TestExport_Driver(t *testing.T) {
	// two baselines × requirements fan out to one line per requirement, in order,
	// each a canonical (key-sorted, newline-terminated) EncodeLine.
	input := []byte(`{"timestamp":"2024-01-01T00:00:00Z","baselines":[` +
		`{"name":"b1","requirements":[{"id":"A"},{"id":"B"}]},` +
		`{"name":"b2","requirements":[{"id":"C"}]}]}`)
	out, err := Export(input, "test-exporter",
		func(req, baseline map[string]interface{}, docTimestamp string, tool, generator, component map[string]interface{}) map[string]interface{} {
			return map[string]interface{}{"id": GetStr(req, "id"), "baseline": GetStr(baseline, "name"), "ts": docTimestamp}
		})
	require.NoError(t, err)
	assert.Equal(t,
		"{\"baseline\":\"b1\",\"id\":\"A\",\"ts\":\"2024-01-01T00:00:00Z\"}\n"+
			"{\"baseline\":\"b1\",\"id\":\"B\",\"ts\":\"2024-01-01T00:00:00Z\"}\n"+
			"{\"baseline\":\"b2\",\"id\":\"C\",\"ts\":\"2024-01-01T00:00:00Z\"}\n",
		string(out))

	// doc-level context is threaded to the builder
	ctxSeen := false
	_, err = Export([]byte(`{"tool":{"name":"t"},"components":[{"name":"h"}],"baselines":[{"requirements":[{"id":"X"}]}]}`), "x",
		func(_, _ map[string]interface{}, _ string, tool, _, component map[string]interface{}) map[string]interface{} {
			assert.Equal(t, "t", GetStr(tool, "name"))
			assert.Equal(t, "h", GetStr(component, "name"))
			ctxSeen = true
			return map[string]interface{}{}
		})
	require.NoError(t, err)
	assert.True(t, ctxSeen)

	noop := func(_, _ map[string]interface{}, _ string, _, _, _ map[string]interface{}) map[string]interface{} {
		return nil
	}
	_, err = Export([]byte(""), "x", noop)
	assert.Error(t, err, "empty input")
	_, err = Export([]byte("not json"), "x", noop)
	assert.Error(t, err, "invalid JSON")
	_, err = Export([]byte(`{"foo":1}`), "x", noop)
	assert.Error(t, err, "missing baselines")
	// empty baselines -> empty output, no error
	out, err = Export([]byte(`{"baselines":[]}`), "x", noop)
	require.NoError(t, err)
	assert.Empty(t, out)
}

func TestFirstCVE(t *testing.T) {
	list := []interface{}{
		map[string]interface{}{"source": "GHSA-xxxx"},
		map[string]interface{}{"source": "cve-2024-1234"}, // case-insensitive prefix
	}
	assert.Equal(t, "cve-2024-1234", FirstCVE(list))
	assert.Equal(t, "", FirstCVE([]interface{}{map[string]interface{}{"source": "NOTCVE"}}))
	assert.Equal(t, "", FirstCVE([]interface{}{map[string]interface{}{"source": "CVE"}})) // too short
	assert.Equal(t, "", FirstCVE(nil))
}

func TestEpochHelpers(t *testing.T) {
	sec, ok := EpochSeconds("2024-01-01T00:00:00Z")
	require.True(t, ok)
	assert.Equal(t, int64(1704067200), sec)
	_, ok = EpochSeconds("")
	assert.False(t, ok)

	ms, ok := EpochMillis("2024-01-01T00:00:00Z")
	require.True(t, ok)
	assert.Equal(t, int64(1704067200000), ms)
	_, ok = EpochMillis("")
	assert.False(t, ok)
}

func TestEncodeLine_NonBMPKeyOrder(t *testing.T) {
	// Keys sort bytewise (UTF-8 == code-point): a < U+E000 < U+1F600. The TS
	// canonicalize must reproduce this exact order (see exportmap.test.ts) — a
	// supplementary-plane key is where JS's default UTF-16 sort diverges.
	line, err := EncodeLine(map[string]interface{}{"a": 1, "": 2, "\U0001F600": 3})
	require.NoError(t, err)
	assert.Equal(t, "{\"a\":1,\"\":2,\"\U0001F600\":3}\n", string(line))
}

func TestEncodeLine_JSONPEscaping(t *testing.T) {
	// Go escapes U+2028/U+2029 in string values (JSONP safety); TS stringifyLine
	// mirrors this so the two stay byte-identical.
	line, err := EncodeLine(map[string]interface{}{"k": "line sep end"})
	require.NoError(t, err)
	assert.Equal(t, "{\"k\":\"line\\u2028sep\\u2029end\"}\n", string(line))
}

func TestBuildHDFBlock(t *testing.T) {
	req := map[string]interface{}{
		"id":               "V-1",
		"effectiveStatus":  "passed",
		"effectiveImpact":  0.3,
		"impact":           0.7,
		"severity":         "high",
		"disposition":      "waiver",
		"tags":             map[string]interface{}{"nist": []interface{}{"AC-6"}, "cci": []interface{}{"CCI-1"}},
		"cvss":             []interface{}{map[string]interface{}{"baseScore": 7.5}},
		"cwe":              []interface{}{"CWE-79"},
		"epss":             0.1,
		"kev":              true,
		"affectedPackages": []interface{}{"pkg"},
		"descriptions":     []interface{}{map[string]interface{}{"label": "default", "data": "d"}},
		"results":          []interface{}{map[string]interface{}{"status": "failed"}},
		"statusOverrides":  []interface{}{map[string]interface{}{"type": "waiver"}},
		"poams":            []interface{}{map[string]interface{}{"id": "P-1"}},
		"code":             "control code",
		"refs":             []interface{}{map[string]interface{}{"url": "https://x"}},
	}
	baseline := map[string]interface{}{"name": "b1"}
	gen := map[string]interface{}{"name": "g"}
	tool := map[string]interface{}{"name": "t"}
	hdf := BuildHDFBlock(req, baseline, "failed", true, true, gen, tool, "0.1.0")

	assert.Equal(t, "failed", hdf["status"])
	assert.Equal(t, true, hdf["overridden"])
	assert.Equal(t, true, hdf["suppressed"])
	assert.Equal(t, "0.1.0", hdf["exporter_version"])
	assert.Equal(t, "V-1", hdf["control_id"])
	assert.Equal(t, "b1", hdf["baseline"])
	assert.Equal(t, "passed", hdf["effective_status"])
	assert.Equal(t, 0.3, hdf["effective_impact"])
	assert.Equal(t, 0.7, hdf["impact"])
	assert.Equal(t, "high", hdf["severity"])
	assert.Equal(t, "waiver", hdf["disposition"])
	assert.Equal(t, []interface{}{"AC-6"}, hdf["nist"])
	assert.Equal(t, []interface{}{"CCI-1"}, hdf["cci"])
	for _, k := range []string{"tags", "cvss", "cwe", "epss", "kev", "affected_packages", "descriptions", "results", "status_overrides", "poams", "refs"} {
		assert.Contains(t, hdf, k, "passthrough key %q present", k)
	}
	assert.Equal(t, "control code", hdf["code"])
	assert.Equal(t, gen, hdf["generator"])
	assert.Equal(t, tool, hdf["tool"])

	// minimal req: optionals + generator/tool absent, empty id skipped by SetIf
	minimal := BuildHDFBlock(map[string]interface{}{}, map[string]interface{}{}, "passed", false, false, nil, nil, "0.1.0")
	assert.Equal(t, "passed", minimal["status"])
	assert.Equal(t, false, minimal["suppressed"])
	assert.NotContains(t, minimal, "effective_status")
	assert.NotContains(t, minimal, "generator")
	assert.NotContains(t, minimal, "tool")
	assert.NotContains(t, minimal, "control_id")
	assert.NotContains(t, minimal, "nist")
}

File: hdf-converters/shared/typescript/exportmap.test.ts
import { describe, it, expect } from 'vitest';
import {
  asMap,
  asArr,
  getStr,
  setIf,
  stringSlice,
  worstOfResults,
  statusOf,
  isFailing,
  firstComponent,
  firstResultStartTime,
  defaultDescription,
  firstRefURL,
  eventID,
  canonicalize,
  stringifyLine,
  runExport,
  firstCVE,
  epochSeconds,
  epochMillis,
  floatNumber,
} from './exportmap.js';

function mkResults(...statuses: string[]): Record<string, unknown> {
  return { results: statuses.map((s) => ({ status: s })) };
}

describe('exportmap status roll-up', () => {
  it('worstOfResults picks the most-significant status', () => {
    expect(worstOfResults(mkResults('passed', 'failed'))).toBe('failed');
    expect(worstOfResults(mkResults('passed', 'passed'))).toBe('passed');
    expect(worstOfResults(mkResults('error', 'passed'))).toBe('error');
    expect(worstOfResults(mkResults('notApplicable'))).toBe('notApplicable');
    expect(worstOfResults(mkResults())).toBe('notReviewed');
  });

  it('statusOf resolves rollup + overridden + suppressed', () => {
    let st = statusOf(mkResults('failed'));
    expect(st).toEqual({ raw: 'failed', effective: '', rollup: 'failed', overridden: false, suppressed: false });

    st = statusOf({ ...mkResults('failed'), effectiveStatus: 'passed' });
    expect(st.raw).toBe('failed'); // lossless roll-up preserved
    expect(st.effective).toBe('passed');
    expect(st.rollup).toBe('passed');
    expect(st.overridden).toBe(true);
    expect(st.suppressed).toBe(true); // raw-failing driven non-failing

    // override present but no effective status: not suppressed (conservative)
    st = statusOf({ ...mkResults('failed'), statusOverrides: [{ type: 'waiver' }] });
    expect(st.overridden).toBe(true);
    expect(st.rollup).toBe('failed');
    expect(st.suppressed).toBe(false);
  });

  it('suppressed axis: waiver/FP suppress, riskAdjustment stays actionable', () => {
    for (const eff of ['passed', 'notApplicable']) {
      expect(statusOf({ ...mkResults('failed'), effectiveStatus: eff }).suppressed).toBe(true);
    }
    // riskAdjustment: effectiveStatus stays failed → still actionable
    const ra = statusOf({
      ...mkResults('failed'),
      effectiveStatus: 'failed',
      statusOverrides: [{ type: 'riskAdjustment', impact: { value: 0.2 } }],
    });
    expect(ra.overridden).toBe(true);
    expect(ra.suppressed).toBe(false);
    // passing / errored findings are never suppressed
    expect(statusOf({ ...mkResults('passed'), effectiveStatus: 'passed' }).suppressed).toBe(false);
    expect(statusOf({ ...mkResults('error'), effectiveStatus: 'passed' }).suppressed).toBe(false);
    expect(isFailing('failed')).toBe(true);
    expect(isFailing('error')).toBe(false);
    expect(isFailing('passed')).toBe(false);
  });
});

describe('exportmap generic access', () => {
  it('asMap / asArr distinguish shapes', () => {
    expect(asMap({ a: 1 })).toEqual({ a: 1 });
    expect(asMap([1])).toBeUndefined();
    expect(asArr([1, 2])).toHaveLength(2);
    expect(asArr({})).toBeUndefined();
  });

  it('getStr returns "" for non-string / missing / undefined map', () => {
    expect(getStr({ a: 'x' }, 'a')).toBe('x');
    expect(getStr({ a: 1 }, 'a')).toBe('');
    expect(getStr(undefined, 'a')).toBe('');
  });

  it('setIf skips empty values', () => {
    const m: Record<string, unknown> = {};
    setIf(m, 'keep', 'v');
    setIf(m, 'drop', '');
    expect(m).toEqual({ keep: 'v' });
  });

  it('stringSlice coerces string or string[]', () => {
    expect(stringSlice('x')).toEqual(['x']);
    expect(stringSlice(['a', 1, 'b'])).toEqual(['a', 'b']);
    expect(stringSlice(42)).toEqual([]);
  });
});

describe('exportmap extraction', () => {
  it('pulls first component / start time / default description / ref url', () => {
    const doc = { components: [{ name: 'web' }] };
    expect(getStr(firstComponent(doc), 'name')).toBe('web');
    expect(firstComponent({})).toBeUndefined();

    const req = {
      results: [{ startTime: '2024-01-01T00:00:00Z' }],
      descriptions: [{ label: 'default', data: 'desc' }],
      refs: [{ url: 'https://x' }],
    };
    expect(firstResultStartTime(req, 'fb')).toBe('2024-01-01T00:00:00Z');
    expect(firstResultStartTime({}, 'fb')).toBe('fb');
    expect(defaultDescription(req)).toBe('desc');
    expect(firstRefURL(req)).toBe('https://x');
  });

  it('eventID joins component | baseline | control', () => {
    expect(eventID({ componentId: 'c1' }, 'base', 'V-1')).toBe('c1|base|V-1');
    expect(eventID(undefined, 'base', 'V-1')).toBe('|base|V-1');
  });
});

describe('exportmap canonical serialization', () => {
  it('canonicalize sorts keys deeply and stringifyLine leaves & unescaped', () => {
    const out = stringifyLine(canonicalize({ b: 1, a: { d: 2, c: 'x&y' } }));
    expect(out).toBe('{"a":{"c":"x&y","d":2},"b":1}');
  });
});

describe('exportmap shared driver + helpers', () => {
  it('runExport fans out one canonical line per requirement, in order', () => {
    const input = JSON.stringify({
      timestamp: '2024-01-01T00:00:00Z',
      baselines: [
        { name: 'b1', requirements: [{ id: 'A' }, { id: 'B' }] },
        { name: 'b2', requirements: [{ id: 'C' }] },
      ],
    });
    const out = runExport(input, 'test-exporter', (req, baseline, docTimestamp) => ({
      id: getStr(req, 'id'),
      baseline: getStr(baseline, 'name'),
      ts: docTimestamp,
    }));
    expect(out).toBe(
      '{"baseline":"b1","id":"A","ts":"2024-01-01T00:00:00Z"}\n' +
        '{"baseline":"b1","id":"B","ts":"2024-01-01T00:00:00Z"}\n' +
        '{"baseline":"b2","id":"C","ts":"2024-01-01T00:00:00Z"}\n',
    );
  });

  it('runExport threads doc-level context to the builder', () => {
    let seen = false;
    runExport(
      JSON.stringify({ tool: { name: 't' }, components: [{ name: 'h' }], baselines: [{ requirements: [{ id: 'X' }] }] }),
      'x',
      (_req, _baseline, _ts, tool, _generator, component) => {
        expect(getStr(tool, 'name')).toBe('t');
        expect(getStr(component, 'name')).toBe('h');
        seen = true;
        return {};
      },
    );
    expect(seen).toBe(true);
  });

  it('runExport throws on invalid/structureless input and returns "" for empty baselines', () => {
    const noop = () => ({});
    expect(() => runExport('', 'x', noop)).toThrow();
    expect(() => runExport('not json', 'x', noop)).toThrow();
    expect(() => runExport('{"foo":1}', 'x', noop)).toThrow(/missing baselines/);
    expect(runExport('{"baselines":[]}', 'x', noop)).toBe('');
  });

  it('firstCVE finds the first CVE-prefixed source (case-insensitive)', () => {
    expect(firstCVE([{ source: 'GHSA-xxxx' }, { source: 'cve-2024-1234' }])).toBe('cve-2024-1234');
    expect(firstCVE([{ source: 'NOTCVE' }])).toBe('');
    expect(firstCVE([])).toBe('');
  });

  it('epoch helpers scale the canonical parse, undefined when unparseable', () => {
    expect(epochSeconds('2024-01-01T00:00:00Z')).toBe(1704067200);
    expect(epochSeconds('')).toBeUndefined();
    expect(epochMillis('2024-01-01T00:00:00Z')).toBe(1704067200000);
    expect(epochMillis('')).toBeUndefined();
  });
});

describe('exportmap non-BMP key parity + JSONP escaping (Go byte-for-byte)', () => {
  // Go's encoding/json sorts object keys bytewise (UTF-8), which equals
  // code-point order. A supplementary-plane key (U+1F600) has a UTF-16 lead
  // surrogate (0xD83D) that sorts BEFORE a BMP key in [0xE000,0xFFFF] under
  // JS's default .sort(), but AFTER it under UTF-8/code-point order. canonicalize
  // must match Go: expected order is a < U+E000 < U+1F600.
  it('sorts non-BMP keys in Go/UTF-8 (code-point) order, not UTF-16', () => {
    const obj = { a: 1, '': 2, '\u{1F600}': 3 };
    expect(stringifyLine(canonicalize(obj))).toBe('{"a":1,"":2,"\u{1F600}":3}');
  });

  it('escapes U+2028/U+2029 in values identically to Go', () => {
    const obj = { k: 'line sep end' };
    expect(stringifyLine(canonicalize(obj))).toBe('{"k":"line\\u2028sep\\u2029end"}');
  });
});

describe('exportmap RawNumber float tokens (OCSF float_t)', () => {
  it('floatNumber renders a bare decimal token, byte-identical to Go json.Number', () => {
    expect(stringifyLine({ base_score: floatNumber(0) })).toBe('{"base_score":0.0}');
    expect(stringifyLine({ base_score: floatNumber(7.5) })).toBe('{"base_score":7.5}');
    expect(stringifyLine({ base_score: floatNumber(8) })).toBe('{"base_score":8.0}');
    expect(stringifyLine({ base_score: floatNumber(10) })).toBe('{"base_score":10.0}');
    // canonicalize preserves the wrapper (does not treat it as an object to sort)
    expect(stringifyLine(canonicalize({ base_score: floatNumber(8) }))).toBe('{"base_score":8.0}');
    // a normal string containing digits is untouched (marker cannot collide)
    expect(stringifyLine({ desc: 'score 10 ok' })).toBe('{"desc":"score 10 ok"}');
  });
});
